lib/zipkin-tracer/rack/zipkin-tracer.rb

```diff
@@ -25,59 +27,33 @@ class RackHandler
     def initialize(app, config = nil)
       @app = app
       @lock = Mutex.new
-
-      config = Config.new(app, config)
-      SuckerPunch.logger = config.logger
-      adapter = config.adapter
-
-      ::Trace.tracer = case adapter
-        when :json
-          require 'zipkin-tracer/zipkin_json_tracer'
-          ::Trace::ZipkinJsonTracer.new(json_api_host: config.json_api_host, traces_buffer: config.traces_buffer)
-        when :scribe
-          require 'zipkin-tracer/careless_scribe'
-          ::Trace::ZipkinTracer.new(CarelessScribe.new(config.scribe_server), config.scribe_max_buffer)
-        when :kafka
-          require 'zipkin-tracer/zipkin_kafka_tracer'
-          ::Trace::ZipkinKafkaTracer.new(zookeepers: config.zookeeper)
-        else
-          ::Trace::NullTracer.new
-      end
-
-      ip_format = config.adapter == :json ? :string : :i32
-      ::Trace.default_endpoint = ::Trace::Endpoint.make_endpoint(
-        nil, # auto detect hostname
-        config.service_port,
-        service_name(config.service_name),
-        ip_format
-      )
-      ::Trace.sample_rate = config.sample_rate
-
-      @config = config
+      @config = Config.new(app, config).freeze
+      @tracer = TracerFactory.new.tracer(@config)
```

Comment:
how about having `attr_reader` for those 2 so you don't manipulate the instance var directly?